pywhy_stats/categorical.py

```diff
@@ -0,0 +1,293 @@
+    Y = np.array(Y)
+    Z = np.array(Z)
+
+    unique_Z_vals = np.unique(Z)
+
+    chi = 0
+    dof = 0
+
+    for z_state in unique_Z_vals:
+        z_indices = np.where(Z == z_state)[0]
+        sub_df = df[z_indices]
+
+        x_values = np.unique(sub_df[X])
+        y_values = np.unique(sub_df[Y])
+
+        counts, _, _ = np.histogram2d(sub_df[X], sub_df[Y], bins=(x_values, y_values))
+
+        if lambda_ is not None:
+            counts = lambda_(counts)
+
+        c, _ = stats.chisquare(counts)
+        chi += c
+        dof += (counts.shape[0] - 1) * (counts.shape[1] - 1)
+
+    return chi, dof
+
+
+# This is a modified function taken from pgmpy: License MIT
+def _power_divergence(
+    X: ArrayLike, Y: ArrayLike, Z: Optional[ArrayLike], lambda_: str = "cressie-read"
+) -> PValueResult:
+    """Compute the Cressie-Read power divergence statistic.
+
+    The null hypothesis for the test is X is independent of Y given Z. A lot of the
+    frequency comparison based statistics (eg. chi-square, G-test etc) belong to
+    power divergence family, and are special cases of this test.
+
+    Parameters
+    ----------
+    X: ArrayLike of shape (n_samples,)
+        The first node variable.
+    Y : ArrayLike of shape (n_samples,)
+        The second node variable.
+    Z : ArrayLike of shape (n_samples, n_variables)
+        The conditioning set.
+    lambda_: float or string
+        The lambda parameter for the power_divergence statistic. Some values of
+        lambda_ results in other well known tests:
+            "pearson"             1          "Chi-squared test"
+            "log-likelihood"      0          "G-test or log-likelihood"
+            "freeman-tukey"     -1/2        "freeman-tukey Statistic"
+            "mod-log-likelihood"  -1         "Modified Log-likelihood"
+            "neyman"              -2         "Neyman's statistic"
+            "cressie-read"        2/3        "The value recommended in the paper
+                                             :footcite:`cressieread1984`"
+
+    Returns
+    -------
+    CI Test Results: tuple
+        Returns a tuple (chi, p_value, dof). `chi` is the
+        chi-squared test statistic. The `p_value` for the test, i.e. the
+        probability of observing the computed chi-square statistic (or an even
+        higher value), given the null hypothesis that X \u27C2 Y | Zs is True.
+        If boolean = True, returns True if the p_value of the test is greater
+        than `significance_level` else returns False.
+
+    See Also
+    --------
+    scipy.stats.power_divergence
+
+    References
+    ----------
+    .. footbibliography::
+    """
+    if X.ndim != 1 or Y.ndim != 1:
+        raise ValueError("X and Y should be 1-D arrays.")
+
+    # ensure we use numpy arrays
+    X = np.asarray(X)
+    Y = np.asarray(Y)
+
+    # Check if all elements are integers
+    if not np.issubdtype(X.dtype, np.integer):
+        le = LabelEncoder()
+        X = le.fit_transform(X)
```

Comment:
- We do not need any order here, right? So, "apple" and "pear" would work as inputs?
- What happen if the inputs are continuous floats? Is the idea to catch this in line 202? But then, this would still go through with a small data set (e.g. 30 samples) since then we have 60 unique values. What about throwing an error when one of the variable is a non-discrete or non-categorical value. There are also checks for this.

Reply:
Idk if there's an agreed upon method for checking if data is categorical? There's a lot of different types possible: https://numpy.org/doc/stable/reference/arrays.dtypes.html. 

I suppose we can do the naive thing and just check for strings/ints and error out otherwise. But my guess is things like `uint` will fail even though it should work, etc.

Reply:
I think to ensure a proper sanity check (we can apply this in the KCI test too), we need to basically have some sort of internal contract of what we expect the input arrays to be once they reach the internal computation (e.g. np.int if categorical). I implemented a more complex preprocessing that does a bunch of error checks. 

At the end it will guarantee:

- each element in each array has the same type per column. Otherwise, what does `['test', 1, 1.0, -1, 'other string']` mean? We can document in the docstring that strings are auto-encoded, but otherwise it most likely is some custom preprocessing that the user needs to take.
- if a column of any (X, Y, Z) array is comprised of string subtypes, it will be encoded using a label encoder
- otherwise if it is a type we do not recognize, we throw an error

Reply:
I believe we don't need this for KCI, as we can adjust the kernel type using a straightforward convention: If it's categorical, encode it as a string. If it's discrete with an order, use integers, and if it's continuous, use floats. In the KCI context, this typically boils down to "if the values in the column are strings, use the delta kernel. Otherwise, default to RBF."

A common point of confusion is that integers or discrete values are not necessarily categorical. Hence, we need an easy way to identify categorical values without requiring the user to know about a specific numpy data type. Given that in many machine learning problems categorical data is already encoded as strings (such as classes, attributes, etc.), simply using the string convention requires the least effort from a user. Most of the time, they don't even need to be aware of this convention.

Regarding `['test', 1, 1.0, -1, 'other string']`, I was thinking about this as well before, but I feel like this is a typical case of "over-engineering", where we don't need to catch all kind of inputs or stuff that can go wrong. Only the less obvious things.

Reply:
Okay that sounds good